storage/local/persistence.go

```diff
@@ -559,13 +607,15 @@ func (p *persistence) loadChunkDescs(fp model.Fingerprint, offsetFromEnd int) ([
 //
 func (p *persistence) checkpointSeriesMapAndHeads(fingerprintToSeries *seriesMap, fpLocker *fingerprintLocker) (err error) {
 	log.Info("Checkpointing in-memory metrics and chunks...")
+	p.checkpointing.Set(1)
 	begin := time.Now()
 	f, err := os.OpenFile(p.headsTempFileName(), os.O_WRONLY|os.O_TRUNC|os.O_CREATE, 0640)
 	if err != nil {
 		return err
 	}
 
 	defer func() {
+		defer p.checkpointing.Set(0)
```

Comment:
Move this just after line 610. Defer's are stacked, so it will then be executed last but in a more obvious fashion than now.